### PR TITLE
Added deepcpy functionality and small test for roadrunner.

### PR DESCRIPTION
**Pull request by PaulJonasJost**
Deepcopy functionality is needed for profiling.

### CODE REVIEW

File: pypesto/objective/roadrunner/road_runner.py [hunk 1]
```diff
@@ -69,6 +70,26 @@ def __init__(
         self.solver_options = solver_options
         super().__init__(x_names=x_names)
 
+    def __deepcopy__(
+        self, memodict: Optional[dict] = None
+    ) -> "RoadRunnerObjective":
+        """Deepcopy function for RoadRunner objective.
+
+        Needed for e.g. profiling.
+        """
```
Comment (dweindl) [inline pypesto/objective/roadrunner/road_runner.py:79]:
```suggestion
        """Deepcopy function for RoadRunner objective."""
```

File: pypesto/objective/roadrunner/road_runner.py [hunk 2]
```diff
@@ -69,6 +70,26 @@ def __init__(
...
+        import roadrunner
+
+        other = self.__class__.__new__(self.__class__)
+
+        for key in set(self.__dict__.keys()) - {"roadrunner_instance"}:
+            other.__dict__[key] = copy.deepcopy(self.__dict__[key])
+        other_rr = roadrunner.RoadRunner()
+        state = self.roadrunner_instance.saveStateS()
+        other_rr.loadStateS(state=state)
+        other.roadrunner_instance = other_rr
```
Comment (dweindl) [inline pypesto/objective/roadrunner/road_runner.py:89]: Maybe you can check with the RR devs, if `__deepcopy__` could be provided by RR.

**Review by dweindl — APPROVED** [review]

File: pypesto/objective/roadrunner/road_runner.py (hunk 2 above)
Comment (dilpath) [reply pypesto/objective/roadrunner/road_runner.py:86]: also can use their `__setstate__` and `__getstate__`

https://github.com/sys-bio/roadrunner/blob/92d063b8877d121c2c1f878fcc1b5ee093b7ec0b/wrappers/Python/roadrunner/roadrunner.i#L1386-L1392

Comment (dweindl) [reply pypesto/objective/roadrunner/road_runner.py:87]: Actually, if we remove

https://github.com/ICB-DCM/pyPESTO/blob/321ee9dbd0223da0c08172fc08d61796e0d176aa/pypesto/objective/base.py#L59-L64

this `__deepcopy__` function here wouldn't be needed. 

Does anybody recall the purpose of `ObjectiveBase.__deepcopy__`?

Comment (dilpath) [reply pypesto/objective/roadrunner/road_runner.py:87]: There's an option that `deepcopy`s the objective at `Problem` instantiation.
https://github.com/ICB-DCM/pyPESTO/blob/34e89b3bc88d2052ca808da43c11c57da75bec04/pypesto/problem/base.py#L107-L108

But I'm not sure why `__deepcopy__` is defined, I would expect a default `deepcopy(...)` to do what is defined in `ObjectiveBase.__deepcopy__` already.

Comment (dweindl) [reply pypesto/objective/roadrunner/road_runner.py:87]: > I would expect a default `deepcopy(...)` to do what is defined in `ObjectiveBase.__deepcopy__` already.

Exactly.

Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/road_runner.py:87]: agreed. Will remove the default deep copy and see whether all tests still work, then merge for now. Might be reverted when RR developers add a deepcopy function.

Comment (dweindl) [reply pypesto/objective/roadrunner/road_runner.py:87]: > Will remove the default deep copy and see whether all tests still work, then merge for now.

Not only `ObjectiveBase.__deepcopy__`, but also `RoadRunnerObjective.__deepcopy__`.

> Might be reverted when RR developers add a deepcopy function.

No, that wouldn't require any change then.

Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/road_runner.py:87]: but RR does not have a deep copy functionality yet afaik. 🤔

Comment (dweindl) [reply pypesto/objective/roadrunner/road_runner.py:87]: `deepcopy` will fall back to `__getstate__`/ `__setstate__`, which exist (see Dilan's comment above)

Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/road_runner.py:87]: good 👍🏼 will remove it but keep the tests, as I think we should still make sure it is working?

File: test/base/test_roadrunner.py [hunk 3]
```diff
@@ -104,3 +106,33 @@ def _execute_case_rr(case, model_type, version):
         )
 
     logger.info(f"Case {version}/{model_type}/{case} passed.")
+
+
+def test_deepcopy():
+    "Test that deepcopy works as intended"
```
Comment (dweindl) [inline test/base/test_roadrunner.py:112]:
```suggestion
    """Test that deepcopy works as intended"""
```

**Pull request merged**